File: nodejs/lesson-02-functions/exercises/exercise.js
'use strict';

// Exercise 1: Write an arrow function `clamp(n, min, max)` that
// returns n if it's in [min, max], min if n < min, max if n > max.
const clamp = (n, min, max) => {
  // TODO: implement (one-line preferred)
  return n;
};

// Exercise 2: Write a function `makeMultiplier(factor)` that returns
// a closure that multiplies its argument by factor.
// const triple = makeMultiplier(3);
// triple(5) → 15
function makeMultiplier(factor) {
  // TODO: implement
  return (n) => n;
}

// Exercise 3: Write `once(fn)` — a higher-order function that returns
// a new function that only executes fn on the FIRST call;
// subsequent calls return the first result without calling fn again.
function once(fn) {
  // TODO: implement using a closure
  return fn;
}

// Exercise 4: Implement a `pipe(...fns)` function that takes any number
// of single-argument functions and returns a new function that passes
// its input through each function left-to-right.
// pipe(double, addOne, square)(3) → (3*2+1)^2 = 49
function pipe(...fns) {
  // TODO: implement using reduce
  return (x) => x;
}

// ─── Main ─────────────────────────────────────────────────────────────────────

console.log('=== Exercise 1: clamp ===');
console.log(clamp(5, 0, 10));   // 5
console.log(clamp(-5, 0, 10));  // 0
console.log(clamp(15, 0, 10));  // 10

console.log('\n=== Exercise 2: makeMultiplier ===');
const triple = makeMultiplier(3);
const double = makeMultiplier(2);
console.log('triple(5):', triple(5));   // 15
console.log('double(7):', double(7));   // 14

console.log('\n=== Exercise 3: once ===');
let callCount = 0;
const init = once(() => {
  callCount++;
  return 'initialized';
});
console.log(init());  // 'initialized'
console.log(init());  // 'initialized' (same result, fn NOT called again)
console.log(init());  // 'initialized'
console.log('fn called:', callCount, 'time(s)'); // should be 1

console.log('\n=== Exercise 4: pipe ===');
const addOne = n => n + 1;
const square = n => n * n;
const toString = n => `result: ${n}`;

const transform = pipe(double, addOne, square, toString);
console.log(transform(3)); // result: 49

File: go/lesson-02-functions/exercises/exercise.go
package main

import "fmt"

// Exercise 1: Write a function that returns both the quotient and remainder
// of integer division. Return an error if divisor is 0.
// divmod(10, 3) → (3, 1, nil)
// divmod(10, 0) → (0, 0, error)
func divmod(a, b int) (int, int, error) {
	// TODO: implement
	return 0, 0, nil
}

// Exercise 2: Write a variadic function that returns the min and max
// of the provided integers. Return an error if no arguments given.
func minMax(nums ...int) (min, max int, err error) {
	// TODO: implement
	return 0, 0, nil
}

// Exercise 3: Write a function that returns a "rate limiter" closure.
// The closure tracks how many times it has been called and returns false
// once the limit is exceeded.
// Example: limit3 := makeLimiter(3)
//   limit3() → true (call 1)
//   limit3() → true (call 2)
//   limit3() → true (call 3)
//   limit3() → false (call 4, over limit)
func makeLimiter(limit int) func() bool {
	// TODO: implement using a closure
	return func() bool { return false }
}

// Exercise 4: Implement a pipeline function that applies a list of
// transformation functions to an integer in sequence.
// pipeline(5, double, addOne, square) applies double → addOne → square
func pipeline(n int, fns ...func(int) int) int {
	// TODO: implement
	return n
}

func main() {
	fmt.Println("=== Exercise 1: divmod ===")
	q, r, err := divmod(10, 3)
	fmt.Printf("10 / 3 = %d remainder %d (err: %v)\n", q, r, err)
	_, _, err = divmod(10, 0)
	fmt.Println("10 / 0 error:", err)

	fmt.Println("\n=== Exercise 2: minMax ===")
	min, max, err := minMax(3, 1, 4, 1, 5, 9, 2, 6)
	fmt.Printf("min=%d max=%d err=%v\n", min, max, err)
	_, _, err = minMax()
	fmt.Println("empty error:", err)

	fmt.Println("\n=== Exercise 3: makeLimiter ===")
	limit3 := makeLimiter(3)
	for i := 0; i < 5; i++ {
		fmt.Printf("call %d: %v\n", i+1, limit3())
	}

	fmt.Println("\n=== Exercise 4: pipeline ===")
	double := func(n int) int { return n * 2 }
	addOne := func(n int) int { return n + 1 }
	square := func(n int) int { return n * n }
	result := pipeline(3, double, addOne, square) // (3*2+1)^2 = 49
	fmt.Println("pipeline(3, double, addOne, square):", result) // 49
}
